### PR TITLE
Things to the API patch

### PR DESCRIPTION
**Pull request by MrIvanPlays**
This patch adds some things to the API. It adds feature to schedule tasks with ticks, and adds some things to the player interface, also removes 2 non reason deprecations.

### CODE REVIEW

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+@@ -16,19 +16,19 @@ public interface CommandSender
+ 
+     /**
+      * Send a message to this sender.
++     * Waterfall - Deprecation removal. No reason deprecating
+      *
+      * @param message the message to send
+      */
+-    @Deprecated
+     public void sendMessage(String message);
```
Comment (hugmanrique) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:32]: This method is deprecated for valid reasons. Legacy texts are converted to text components, but you should always use the latter.

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+ 
+     /**
+      * Send several messages to this sender. Each message will be sent
+      * separately.
++     * Waterfall - Deprecation removal. No reason deprecating
+      *
+      * @param messages the messages to send
+      */
+-    @Deprecated
```
Comment (hugmanrique) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:41]: Same thing here

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+--- a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
++++ b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
+@@ -69,7 +69,9 @@ public interface ProxiedPlayer extends Connection, CommandSender
+      *
+      * @param position the screen position
+      * @param message the message to send
++     * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} or {@link #sendActionBar(BaseComponent)}, {@link #sendActionBar(BaseComponent...)} for a more proper ActionBar API.
+      */
++    @Deprecated // Waterfall
+     public void sendMessage(ChatMessageType position, BaseComponent... message);
```
Comment (hugmanrique) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:56]: This method is perfectly valid, why would we deprecate it?

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+ 
+     /**
+@@ -77,7 +79,9 @@ public interface ProxiedPlayer extends Connection, CommandSender
+      *
+      * @param position the screen position
+      * @param message the message to send
++     * @deprecated This is unlikely the API you want to use. See {@link #sendActionBar(String)} or {@link #sendActionBar(BaseComponent)}, {@link #sendActionBar(BaseComponent...)} for a more proper ActionBar API.
+      */
++    @Deprecated // Waterfall
+     public void sendMessage(ChatMessageType position, BaseComponent message);
```
Comment (hugmanrique) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:66]: Same thing as above.

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+index 81bd18f0..44ff7cd9 100644
+--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
++++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+@@ -181,10 +181,14 @@ public class PluginManager
+             {
+                 if ( proxy.getConfig().isLogCommands() )
+                 {
+-                    proxy.getLogger().log( Level.INFO, "{0} executed command: /{1}", new Object[]
++                    // Waterfall - check if the command sender is not the console
++                    if ( !sender.getName().equalsIgnoreCase( proxy.getConsole().getName() ) )
```
Comment (hugmanrique) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:135]: Logging console commands might be useful for web terminals/screen/tmux... where multiple people might be watching the same console output

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
++
++    /**
++     * Schedules a task to be executed asynchronously after the specified delay
++     * is up.
++     *
++     * @param owner the plugin owning this task
++     * @param task the task to run
++     * @param delay the delay before this task will be executed, specified in ticks
++     * @return the scheduled task
++     */
```
Comment (hugmanrique) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:165]: Bungee (and by extension, Waterfall) doesn't have a concept of ticks. This is only used in NMS based servers. You can pass 50 millis if you want to run a task every tick.

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
++    @Getter
++    private Title currentTitle;
+     // Waterfall end
+     /*========================================================================*/
+     @Getter
+@@ -709,6 +711,7 @@ public final class UserConnection implements ProxiedPlayer
+     @Override
+     public void sendTitle(Title title)
+     {
++        this.currentTitle = title; // Waterfall
```
Comment (hugmanrique) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:204]: I don't see any current title expiration code

**Review by hugmanrique — CHANGES_REQUESTED** [review]
I like the additions of sendActionBar and sendTitle, but the other stuff would require too much maintenance + it should be in another separate PR (and patch)

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch (hunk 5 above)
Comment (Janmm14) [reply BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:135]: And generally useful when checking the log file?

**Review by Janmm14 — COMMENTED** [review]
Heck, we have the fancy sendMessage with chat position. I don't see these sendActionBar function a useful abbreviation. It might even add further confusion in combination with semdMessage with action bar position.

I'd say waterfall should not encourage alternatives to working api in bungee.

And btw. there are 3 message types. When you deprecate the function with the enum and don't add functions with chat and system type difference, its not good.

**Comment by Janmm14** [comment]
And "Add things to the API" as description is also not ok. What you did here belongs into multiple commits/patches with more descriptive naming. (Apart from most changes I and the previous reviewer think are not good)

**Comment by hugmanrique** [comment]
Yeah, on a second thought, adding an extra method every time a new chat position is created is not scalable.

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+     Scoreboard getScoreboard();
++
++    // Waterfall start
++
++    /**
++     * Sends ActionBar message to the client.
++     *
++     * @param message the message you wish to send
++     */
++    public void sendActionBar(String message);
```
Comment (kashike) [inline BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:81]: Add `@Deprecated` to this method.

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by MrIvanPlays** [comment]
That's why i didn't wanted to PR, but everyone said 'just PR to waterfall'. I KNEW that it will be rejected...

**Comment by aikar** [comment]
You need to be able to take feedback. The quality of your submission was low, and people have provided feedback on what was wrong so that you can learn to improve the quality of your contributions.

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch (hunk 6 above)
Comment (mdcfe) [reply BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:165]: In addition, you should specify either in Javadocs or the method name (or both) that this method accepts ticks. It isn't obvious what unit the delay is in unless you read the implementation code.

File: BungeeCord-Patches/0049-Add-some-things-to-the-API.patch (hunk 1 above)
Comment (MrIvanPlays) [reply BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:32]: Then why md_5 doesn't specify it in javadocs as '@ deprecated' ?!

Comment (mdcfe) [reply BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:32]: You can already see it's deprecated from the annotation on the method; he probably felt it wasn't necessary to duplicate that line.

Comment (aikar) [reply BungeeCord-Patches/0049-Add-some-things-to-the-API.patch:32]: Honestly, Deprecating string methods is stupid, as it serves a purpose to send a simple message.

One could argue the legacy codes support of this message are 'deprecated' and that the method isn't, but it's also possible to permanently support that system on the API, which I'm supportive of, as components are stupid for simple messaging. 

I'm in favor of permanently keeping legacy codes around for simple messaging, and if mojang adds more advanced capabilities than current with components, then the legacy system just wouldn't expose it and people can move to comps then.

But even a sendMessage(ChatColor color, String msg); that doesn't use legacy codes would be useful to declare simple messages, but I see no harm in keeping legacy codes on sendMessage.

If md5 removed it from Bukkit API, I would add it back. It costs nothing to maintain.

**Pull request closed without merging**